---
title: "OICR-9429"
author: "Alessandro Vai"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
library(tidyverse) |> suppressPackageStartupMessages()
library(DESeq2) |> suppressPackageStartupMessages()
```

```{r scaling factor}
# From SpikeFlow output
report <- read.csv("SplitBam_qc_FULL.csv", row.names = 1, check.names = F) |> 
    rownames_to_column("Sample")

scaling_factor_df <- inner_join(report |>
               filter(startsWith(Sample, "IP")) |>
               mutate(Sample = str_remove(Sample, "IP_")),
           report |>
               filter(startsWith(Sample, "input")) |>
               mutate(Sample = str_remove(Sample, "input_")),
           by = "Sample", suffix = c(".IP", ".INPUT")) |> 
           # Paper normalization
           # https://www.sciencedirect.com/science/article/pii/S1535610819301990#sec5
    mutate(scaling_factor = (`Tot. Sample Reads.IP` / `Tot. spikeIn Reads.IP`) / 
                            (`Tot. Sample Reads.INPUT` / `Tot. spikeIn Reads.INPUT`),
           Sample = str_replace(Sample, "-", "."),
           Sample = paste0("IP_", Sample)) |> 
    dplyr::select(Sample, scaling_factor)
```

```{r DESeq}
cts <- read.delim("count231.txt", comment.char = "#") |> 
    dplyr::select(-c(Chr,Start,End,Strand,Length)) |> 
    column_to_rownames("Geneid")
colnames(cts) <- str_remove(colnames(cts), "_ref.sorted.bam")

# DMSO REP3 is out from PCA
cts$IP_DMSO.rep3 <- NULL
coldata <- data.frame(row.names = colnames(cts),
                      condition = str_extract(colnames(cts), "\\_.*\\.") |> 
                      str_remove("\\_") |> str_remove("\\."))

dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ 0 + condition)
    
smallestGroupSize <- 2
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
scaling_factor_df <- scaling_factor_df |>
      filter(!Sample %in% c("IP_DMSO.rep3"))
# Turn Off DESeq scaling
sizeFactors(dds) <- 1 / scaling_factor_df[match(colnames(dds),
                                                scaling_factor_df$Sample), ] |>
    pull(scaling_factor)


dds <- DESeq(dds, quiet = T)

OICRvsDMSO <- results(dds,
                      contrast = c("condition", "OICR", "DMSO"),
                      alpha = 0.05)
```

```{r Fig 6d}
ggplotma <- function(count, result, color_signif) {
    count |> as_tibble(rownames = "PeakID") |> 
    inner_join(result |> as_tibble(rownames = "PeakID")) |> 
    mutate(Legend = if_else(is.na(padj) | padj > 0.05, "NR", "Reg")) |> 
    ggplot(aes(x = value, y = log2FoldChange, color = Legend, alpha = 0.8)) +
    ggrastr::geom_point_rast() + theme_classic() +
    scale_x_log10() +
    geom_hline(yintercept = 0) +
    scale_color_manual(values = c("gray", color_signif)) +
    theme(legend.position = "none",
          text = element_text(size = 14)) +
    xlab("Mean of normalized counts")
}

ggplotma(rowMeans(counts(dds, norm=T)[, c(1,2,5:7)]), OICRvsDMSO, "orchid")
```


```{r peak-to-genes}
# BED file containing coordinates of promoter and TTS of genes
prom3kb <- read.delim("prom3kb+geneBody.bed",
                      header = F, col.names = c("chr", "start", "end",
                                                "Gene")) |> 
  distinct(.keep_all = T)
atac_enh <- read.delim("TCGA-ATAC_DataS7_PeakToGeneLinks_v2_BRCA_Links_PMID30361341.bed",
                       header = F, col.names = c("chr", "start", "end", "Gene"))

prom_enh_gr <- bind_rows(prom3kb, atac_enh) |> 
  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = T)
rm(prom3kb, atac_enh)
OICRvsDMSOgr <- OICRvsDMSO |> as.data.frame() |>
    rownames_to_column("PeakID") |> 
    inner_join(saf) |> 
  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = T)
```


```{r Supp Fig 10b}
overlapped <- findOverlaps(OICRvsDMSOgr, prom_enh_gr)
OICRvsDMSOgr_overlapped <- OICRvsDMSOgr[overlapped@from,]
OICRvsDMSOgr_overlapped$Gene <- prom_enh_gr[overlapped@to, ]$Gene
peak_to_gene <- OICRvsDMSOgr_overlapped |> as.data.frame() |> 
  distinct() |> 
  mutate(isMin = padj == min(padj), .by = Gene) |> 
  filter(Gene %in% c("PLA2G4A", "IGF2BP3", "GLS",
          "PLCG2", "GLIPR2", "AIM2", "TRIM2",
          "EGFR", "MSLN", "PRSS3",
          "SUV39H2", "TMEM71", "TRPV4",
          "GABRE", "ALDH1A3", "CHRM3"),
         isMin)

data <- counts(dds, norm=T) |> 
  as.data.frame() |> as_tibble(rownames="PeakID") |> 
  inner_join(peak_to_gene) |> 
  select(1:6, Gene) |> 
  pivot_longer(!c(PeakID, Gene), names_to = "Sample", values_to = "Count") |> 
  mutate(Treatment = if_else(str_detect(Sample, "DMSO"), "DMSO", "OICR"))

# Calculates mean, sd, se
my_sum <- data |> 
  summarise( 
    n=n(),
    mean=mean(Count),
    sd=sd(Count),
    .by =c(Gene, Treatment)) |> 
  mutate( se=sd/sqrt(n) )


# Standard Error
ggplot(my_sum) +
  geom_bar(aes(x=Gene, y=mean, fill=Treatment, group=Treatment), 
           stat="identity", alpha=0.5, 
           position = position_dodge(width = 1)) +
  geom_errorbar(aes(x=Gene, ymin=mean-se, ymax=mean+se, group=Treatment), 
                width=0.25,
                colour="orange", alpha=0.9, #linewidth=1.5,
                position = position_dodge(width = 1)) +
  theme_classic() +
  scale_fill_manual(values = c("OICR" = "orchid",
                               "DMSO" = "black")) +
  ylab("Normalized count") +
  theme(text = element_text(family = "ArialMT", size = 7),
        axis.text.x = element_text(angle = 45, hjust=1))
```


